### PR TITLE
add sendAjax with custom headers

### PR DESCRIPTION
**Pull request by Sayalic0**
Adding sendAjax with custom headers. See [Issue#1560](https://github.com/casperjs/casperjs/issues/1560).

### CODE REVIEW

**Comment by paazmaya** [comment]
In order to be backwards compatible, check that if the `headers` argument is boolean, then it would in fact be the `async`, etc

**Comment by istr** [comment]
There is already a `settings` argument. No need to change the signature, just pass headers as sub-object / property of the `settings` and check for them.

File: modules/clientutils.js [hunk 1]
```diff
@@ -849,11 +849,12 @@
...
          * @param   String   method   HTTP method (default: GET).
          * @param   Object   data     Request parameters.
+         * @param   Object   headers  Request headers.
          * @param   Boolean  async    Asynchroneous request? (default: false)
          * @param   Object   settings Other settings when perform the ajax request
          * @return  String            Response text.
          */
-        this.sendAJAX = function sendAJAX(url, method, data, async, settings) {
+        this.sendAJAX = function sendAJAX(url, method, data, headers, async, settings) {
             var xhr = new XMLHttpRequest(),
```
Comment (istr) [inline modules/clientutils.js]: Do not change the signature, pass it within settings.

File: modules/clientutils.js [hunk 2]
```diff
@@ -864,6 +865,9 @@
             if (settings && settings.overrideMimeType) {
                 xhr.overrideMimeType(settings.overrideMimeType);
             }
+            for(var header in headers) {
```
Comment (istr) [inline modules/clientutils.js]:
```js
if (settings && settings.headers) {
  for (var header in settings.headers) {
...
```

**Comment by istr** [comment]
@Sayalic0 This PR is a good start, even if you qualify yourself as "a beginner for JavaScript." (#1560). Just integrate the supposed change, add a test case and documentation and this PR is very likely to be accepted.

Thanks for your contribution. :-)

**Comment by Sayalic0** [comment]
Thanks for kind reply of you guys. I will check your advice carefully and refine my PR.

**Comment by Sayalic0** [comment]
@istr I refined my code and wrote docs about the my new feature. Now I get stuck in writing test. I just have no idea about how to write a proper unit test for `sendAJAX`. Should I start a web server in a unit test, then send a request to that web server and check the header?

**Comment by istr** [comment]
@Sayalic0 Thanks for your effort. Please be patient, I do not have the time to give a good advice before next weekend. Thanks for understanding.

**Comment by Sayalic0** [comment]
@istr No hurry. Take your time and thanks for you reply.

**Pull request closed without merging**